### DISCUSSION TITLE
METIS vs R-tree (conditioning)

### DISCUSSION THREAD

**Discussion by fdrmrc** [fdrmrc/non_matching_test_suite#38] [Show and tell]
I'm investigating the conditioning of the two approaches. Based on the results, I can do the following observations:

1. METIS matrix is not singular, but it always has a higher estimated condition number $K(A)$ than the R-tree one,
2. The larger gap in $K(A)$ occurs when partitions are bad, and as expected there we have a kink in the convergence plot,
3. Iteration counts are always sensibly lower for the R-tree.

@andreacangiani @luca-heltai Feel free to write here other possible tests (sorry for spamming)

## EXAMPLE 1: _Structured_ square mesh
- $\Omega= [0,1]^2$
- Size underlying fine tria: 65 536 elements
- Manufactured solution $u=xy(x-1)(y-1)$
- Penalty `sigma = 50*(1/element_diameter)`

Estimated condition number $K(A)$ | CG iterations  
:-------------------------:|:-------------------------:
![Condition_square_structured](https://github.com/fdrmrc/Polydeal/assets/50499769/08e52ca1-734c-41c8-ad5b-fa730db13896) |  ![CGIteration_counts](https://github.com/fdrmrc/Polydeal/assets/50499769/ed24a705-5f0e-471d-adf8-ca121e5a2445)

And this is the associated convergence plot I showed you last week:
![Screenshot 2023-12-19 at 15 25 12](https://github.com/fdrmrc/Polydeal/assets/50499769/f8c5b0d7-70b7-4892-8988-88436e00a3fe)

**Comment by luca-heltai** [comment]
These are very good results. Now I'd really like to see what happens with MG. :)

**Reply by fdrmrc** [reply]
Yes! I need to figure out what the penalty parameter should be before moving on. After several checks, I am convinced that the kink we are observing with METIS is due to a bad choice of $\sigma(x)$. Right now I am working on implementing the following one, which is the one used by Zhaonan: $$\sigma(x) = 10 \frac{(p+1)(p+d)}{h_f^{\perp}}$$


The goal is to have the behaviour we see in the preasymptotic regime also for larger number of DoFs.

**Comment by fdrmrc** [comment]
@andreacangiani @luca-heltai 

Using the penalty: $\sigma(x) = 10 \frac{(p+1)(p+d)}{h_f^{\perp}}$ improves the situtation observed with METIS. Considering again as $\Omega_h$ the perfectly structured square with the configuration as in the message above:

- This is the convergence history with the old penalty `sigma=50/current_element_diameter`
![convergence_structured_classical_penalty](https://github.com/fdrmrc/Polydeal/assets/50499769/7f2e0dfc-7a06-4092-99f2-b01e5c2ab7f6)

- - - - - - - - - - -
- Here's the result after using $\sigma$ defined as above:

![convergence_structured](https://github.com/fdrmrc/Polydeal/assets/50499769/50f8b7ba-442c-49d7-99d2-ee09c579b3d7)

**Reply by fdrmrc** [reply]
By the way, the last mesh produced with METIS (for which we have a kink) is the usual bad one... 


![immagine](https://github.com/fdrmrc/Polydeal/assets/50499769/5defd6d0-8e05-431b-9697-977f91df53d7)

**Comment by luca-heltai** [comment]
This is more credible. And I still like a lot the fact that the error is a factor two better with the r-tree structure. :)

**Comment by andreacangiani** [comment]
This is really good work, thank Marco. 

Just to make sure: you are using _exactly_ the same definition for the penalty for both meshes, right? (I want to make sure you are not over-penalising in the METIS case.)

Did you really manga to compute the orthogonal h_F? (This may vary from one to a few of the small h) WELL DONE!!!

For comparison, this is a result from the first DGP paper (solution being product of sine functions):

<img width="719" alt="Screenshot 2023-12-23 at 11 54 38" src="https://github.com/fdrmrc/Polydeal/assets/55988311/ef7dc8fc-fe20-4e1a-af14-82741913b154">

The bad voronoi mesh, which is not as bad as out METIS mesh, does have also larger errors, although not as larger. Note that you need to compare with p=1 results. But overall it seems we are pretty close to previous observations now. Interesting that there is still a little decrees in rate for the last result. It would be interesting to investigate cases where the geometry does require an even more refined choice of the penalty, but this is not going to be easy and probably out of scope for us. The results are already convincing enough!

It would be good to see H^1-seminorm as well.  And try with p-bases as well... :-o

**Reply by fdrmrc** [reply]
> Just to make sure: you are using exactly the same definition for the penalty for both meshes, right? (I want to make sure you are not over-penalising in the METIS case.)

Yes, exactly the same code, and hence same penalty. I only tweak an "agglomeration strategy" parameter.

> Did you really manage to compute the orthogonal h_F? (This may vary from one to a few of the small h)

Yes, here's how I did. Given a face $f$ of a polygon $P$, to compute $h_f$ I do the following steps:
- shoot a ray from the midpoint $p$ of the face $f$,
- find the point $q$ on the other edge of the polygon that is intersected by such a ray,
- set $h_f = ||p-q||$

If I remember correctly, this is what we discussed, right?

> It would be good to see H^1-seminorm as well. And try with p-bases as well... :-o

I could also try with tensor product Legendre as done in the book

**Reply by fdrmrc** [reply]
Here's a plot with $H^1$-seminorm included for the structured square test. As before, there's a significant gap between METIS and Rtree:

![convergence_structured](https://github.com/fdrmrc/Polydeal/assets/50499769/2fc164ab-d305-4f0d-a5e4-a67267809ce5)
